services/replicator/inconn.go

```diff
@@ -75,69 +73,59 @@ func newInConnection(extUUID string, destPath string, stream storeStream.BStoreO
 		destM3Client:         metrics.NewClientWithTags(m3Client, metrics.Replicator, common.GetDestinationTags(destPath, localLogger)),
 		metricsScope:         metricsScope,
 		perDestMetricsScope:  perDestMetricsScope,
-		closeChannel:         make(chan struct{}),
 		creditsCh:            make(chan int32, 5),
 		creditFlowExpiration: time.Now().Add(creditFlowTimeout),
+		shutdownCh:           make(chan struct{}),
 	}
 
 	return conn
 }
 
 func (conn *inConnection) open() {
-	conn.lk.Lock()
-	defer conn.lk.Unlock()
-
-	if !conn.opened {
-		go conn.writeMsgsStream()
-		go conn.readCreditsStream()
-
-		conn.opened = true
-	}
+	conn.wg.Add(2)
+	go conn.writeMsgsStream()
+	go conn.readCreditsStream()
 	conn.logger.Info("in connection opened")
 }
 
-func (conn *inConnection) close() {
-	conn.lk.Lock()
-	defer conn.lk.Unlock()
+func (conn *inConnection) WaitUntilDone() {
+	conn.wg.Wait()
+}
 
-	if !conn.closed {
-		close(conn.closeChannel)
-		conn.closed = true
-	}
-	conn.logger.Info("in connection closed")
+func (conn *inConnection) shutdown() {
+	close(conn.shutdownCh)
```

Comment:
You might want to put a log line here .. so you know the reasons that the go-routines have gone, etc.